staging/src/k8s.io/component-base/metrics/counter.go

```diff
@@ -107,6 +108,8 @@ func (c *Counter) initializeDeprecatedMetric() {
 
 // WithContext allows the normal Counter metric to pass in context.
 func (c *Counter) WithContext(ctx context.Context) CounterMetric {
+	c.m.Lock()
+	defer c.m.Unlock()
```

Comment:
I believe the change got in due to the fact that the current context carries traits such as span and trace IDs, which would be unique per counter instance. However, the pattern suggested by you is actually what we have been using for vectors, and makes sense, while also bullet-proofing us for the future (when non-unique values may be added to the context as well).

Thank you for catching this, I've refactored the code, PTAL.

Reply:
Ah, seems I need to look more into this, on it.

Reply:
Ah, forgot to add tests, on it.

Reply:
Hmm, correct me if I'm wrong here but I don't think we need tests anymore after the last change, since the reference copies being generated now guarantee no race conditions?

Reply:
A test should verify that the follow-up calls really get the context. This may seem trivial, but I think it was broken before.